File: Practical_03.py
def input_matrix(rows, cols):
    """Function to input a matrix from the user."""
    matrix = []
    print(f"Enter the elements of the matrix ({rows}x{cols}):")
    for i in range(rows):
        row = list(map(int, input(f"Enter row {i + 1}: ").split()))
        if len(row) != cols:
            print(f"Row {i + 1} should have exactly {cols} elements.")
            return None
        matrix.append(row)
    return matrix

def display_matrix(matrix):
    """Function to display a matrix."""
    for row in matrix:
        print(" ".join(map(str, row)))

def add_matrices(matrix1, matrix2):
    """Function to add two matrices."""
    return [[matrix1[i][j] + matrix2[i][j] for j in range(len(matrix1[0]))] for i in range(len(matrix1))]

def subtract_matrices(matrix1, matrix2):
    """Function to subtract two matrices."""
    return [[matrix1[i][j] - matrix2[i][j] for j in range(len(matrix1[0]))] for i in range(len(matrix1))]

def multiply_matrices(matrix1, matrix2):
    """Function to multiply two matrices."""
    result = [[0 for _ in range(len(matrix2[0]))] for _ in range(len(matrix1))]
    for i in range(len(matrix1)):
        for j in range(len(matrix2[0])):
            for k in range(len(matrix2)):
                result[i][j] += matrix1[i][k] * matrix2[k][j]
    return result

def transpose_matrix(matrix):
    """Function to transpose a matrix."""
    return [[matrix[j][i] for j in range(len(matrix))] for i in range(len(matrix[0]))]

# Main function to perform operations
def main():
    print("Matrix Operations Program")
    print("1. Addition of Two Matrices")
    print("2. Subtraction of Two Matrices")
    print("3. Multiplication of Two Matrices")
    print("4. Transpose of a Matrix")
    
    choice = int(input("Choose an operation (1-4): "))

    if choice == 1 or choice == 2:
        rows = int(input("Enter number of rows: "))
        cols = int(input("Enter number of columns: "))
        print("\nMatrix 1:")
        matrix1 = input_matrix(rows, cols)
        print("\nMatrix 2:")
        matrix2 = input_matrix(rows, cols)

        if matrix1 and matrix2:
            if choice == 1:
                print("\nAddition of the two matrices:")
                result = add_matrices(matrix1, matrix2)
            else:
                print("\nSubtraction of the two matrices:")
                result = subtract_matrices(matrix1, matrix2)
            display_matrix(result)

    elif choice == 3:
        rows1 = int(input("Enter number of rows for Matrix 1: "))
        cols1 = int(input("Enter number of columns for Matrix 1: "))
        print("\nMatrix 1:")
        matrix1 = input_matrix(rows1, cols1)

        rows2 = int(input("Enter number of rows for Matrix 2: "))
        cols2 = int(input("Enter number of columns for Matrix 2: "))
        print("\nMatrix 2:")
        matrix2 = input_matrix(rows2, cols2)

        if matrix1 and matrix2:
            if cols1 == rows2:
                print("\nMultiplication of the two matrices:")
                result = multiply_matrices(matrix1, matrix2)
                display_matrix(result)
            else:
                print("Error: Number of columns in Matrix 1 must equal number of rows in Matrix 2 for multiplication.")

    elif choice == 4:
        rows = int(input("Enter number of rows for the matrix: "))
        cols = int(input("Enter number of columns for the matrix: "))
        print("\nMatrix:")
        matrix = input_matrix(rows, cols)

        if matrix:
            print("\nTranspose of the matrix:")
            result = transpose_matrix(matrix)
            display_matrix(result)

    else:
        print("Invalid choice! Please select an option between 1 and 4.")

# Run the main function
main()

File: Practical_03.c
#include <stdio.h>

void inputMatrix(int matrix[10][10], int row, int col) {
    printf("Enter elements of the matrix (%dx%d):\n", row, col);
    for (int i = 0; i < row; i++) {
        for (int j = 0; j < col; j++) {
            printf("Element [%d][%d]: ", i + 1, j + 1);
            scanf("%d", &matrix[i][j]);
        }
    }
}

void printMatrix(int matrix[10][10], int row, int col) {
    for (int i = 0; i < row; i++) {
        for (int j = 0; j < col; j++) {
            printf("%d ", matrix[i][j]);
        }
        printf("\n");
    }
}

// Function to add two matrices
void addMatrices(int mat1[10][10], int mat2[10][10], int result[10][10], int row, int col) {
    for (int i = 0; i < row; i++) {
        for (int j = 0; j < col; j++) {
            result[i][j] = mat1[i][j] + mat2[i][j];
        }
    }
}

// Function to subtract two matrices
void subtractMatrices(int mat1[10][10], int mat2[10][10], int result[10][10], int row, int col) {
    for (int i = 0; i < row; i++) {
        for (int j = 0; j < col; j++) {
            result[i][j] = mat1[i][j] - mat2[i][j];
        }
    }
}

// Function to multiply two matrices
void multiplyMatrices(int mat1[10][10], int mat2[10][10], int result[10][10], int row1, int col1, int col2) {
    for (int i = 0; i < row1; i++) {
        for (int j = 0; j < col2; j++) {
            result[i][j] = 0;
            for (int k = 0; k < col1; k++) {
                result[i][j] += mat1[i][k] * mat2[k][j];
            }
        }
    }
}

// Function to transpose a matrix
void transposeMatrix(int matrix[10][10], int transposed[10][10], int row, int col) {
    for (int i = 0; i < row; i++) {
        for (int j = 0; j < col; j++) {
            transposed[j][i] = matrix[i][j];
        }
    }
}

int main() {
    int mat1[10][10], mat2[10][10], result[10][10];
    int row1, col1, row2, col2;
    int choice;
    int continueProgram = 1;

    while (continueProgram) {
        printf("\n----- Matrix Operations Menu -----\n");
        printf("1. Addition of two matrices\n");
        printf("2. Subtraction of two matrices\n");
        printf("3. Multiplication of two matrices\n");
        printf("4. Transpose of a matrix\n");
        printf("5. Exit\n");
        printf("Enter your choice: ");
        scanf("%d", &choice);

        switch (choice) {
            case 1: // Addition
                printf("Enter the number of rows and columns of the matrices: ");
                scanf("%d%d", &row1, &col1);
                printf("Enter elements of first matrix:\n");
                inputMatrix(mat1, row1, col1);
                printf("Enter elements of second matrix:\n");
                inputMatrix(mat2, row1, col1);
                addMatrices(mat1, mat2, result, row1, col1);
                printf("Result of addition:\n");
                printMatrix(result, row1, col1);
                break;

            case 2: // Subtraction
                printf("Enter the number of rows and columns of the matrices: ");
                scanf("%d%d", &row1, &col1);
                printf("Enter elements of first matrix:\n");
                inputMatrix(mat1, row1, col1);
                printf("Enter elements of second matrix:\n");
                inputMatrix(mat2, row1, col1);
                subtractMatrices(mat1, mat2, result, row1, col1);
                printf("Result of subtraction:\n");
                printMatrix(result, row1, col1);
                break;

            case 3: // Multiplication
                printf("Enter rows and columns of first matrix: ");
                scanf("%d%d", &row1, &col1);
                printf("Enter rows and columns of second matrix: ");
                scanf("%d%d", &row2, &col2);

                if (col1 != row2) {
                    printf("Matrix multiplication not possible.\n");
                    break;
                }

                printf("Enter elements of first matrix:\n");
                inputMatrix(mat1, row1, col1);
                printf("Enter elements of second matrix:\n");
                inputMatrix(mat2, row2, col2);
                multiplyMatrices(mat1, mat2, result, row1, col1, col2);
                printf("Result of multiplication:\n");
                printMatrix(result, row1, col2);
                break;

            case 4: // Transpose
                printf("Enter rows and columns of the matrix: ");
                scanf("%d%d", &row1, &col1);
                printf("Enter elements of the matrix:\n");
                inputMatrix(mat1, row1, col1);
                transposeMatrix(mat1, result, row1, col1);
                printf("Transpose of the matrix:\n");
                printMatrix(result, col1, row1);
                break;

            case 5:
                continueProgram = 0;
                printf("Exiting program.\n");
                break;

            default:
                printf("Invalid choice! Please try again.\n");
                break;
        }
    }
    return 0;
}
